packages/flutter/lib/src/rendering/unconstrained_box.dart

```diff
@@ -0,0 +1,170 @@
+    RenderBox child,
+    @required TextDirection textDirection,
+    @required AlignmentGeometry alignment,
+  }) : assert(alignment != null),
+       _textDirection = textDirection,
+       _alignment = alignment {
+    this.child = child;
+  }
+
+  /// Determines the order to lay children out horizontally and how to interpret
+  /// `start` and `end` in the horizontal direction for the alignment.
+  ///
+  /// The textDirection is only used when [alignment] is an
+  /// [AlignmentDirectional], but must be non-null in that case.
+  @override
+  TextDirection get textDirection => _textDirection;
+  TextDirection _textDirection;
+  set textDirection(TextDirection value) {
+    if (_textDirection != value) {
+      _textDirection = value;
+      markNeedsLayout();
+    }
+  }
+
+  /// How to align the the child in the box.
+  ///
+  /// If this is set to an [AlignmentDirectional] object, then [textDirection]
+  /// must not be null.
+  ///
+  /// See also:
+  ///  * [AlignmentDirectional] for direction-aware alignment.
+  ///  * [Alignment] for non-direction-aware alignment.
+  AlignmentGeometry get alignment => _alignment;
+  AlignmentGeometry _alignment;
+  set alignment(AlignmentGeometry value) {
+    assert(value != null);
+    if (_alignment == value)
+      return;
+    _alignment = value;
+    _markNeedResolution();
+  }
+
+  Alignment _resolvedAlignment;
+  void _resolve() {
+    _resolvedAlignment ??= alignment.resolve(textDirection ?? TextDirection.ltr);
+  }
+
+  void _markNeedResolution() {
+    _resolvedAlignment = null;
+    markNeedsLayout();
+  }
+
+  @override
+  void setupParentData(RenderBox child) {
+    if (child.parentData is! BoxParentData)
+      child.parentData = new BoxParentData();
+  }
+
+  @override
+  void performLayout() {
+    if (child != null) {
+      _resolve();
+      assert(_resolvedAlignment != null);
+      final BoxParentData childParentData = child.parentData;
+      // Let the child lay itself out at it's "natural" size.
+      child.layout(const BoxConstraints(), parentUsesSize: true);
+      size = constraints.constrain(child.size);
+      childParentData.offset = _resolvedAlignment.alongOffset(size - child.size);
+      overflowChildRect = childParentData.offset & child.size;
+      overflowContainerRect = Offset.zero & size;
+    } else {
+      size = constraints.constrain(Size.zero);
+    }
+  }
+
+  @override
+  void paint(PaintingContext context, Offset offset) {
+    // There's no point in drawing the child if we're empty, or there is no
+    // child.
+    if (child == null || size.isEmpty)
+      return;
+
+    final BoxParentData childParentData = child.parentData;
+    if (!isOverflowing) {
+      super.paint(context, offset + childParentData.offset);
+      return;
+    }
+
+    // We have overflow. Clip it.
+    context.pushClipRect(
+      needsCompositing,
+      offset + childParentData.offset,
+      (Offset.zero - childParentData.offset) & size,
+      super.paint,
```

Comment:
that seems like a weird clip rect. why not just `context.pushClipRect(needsCompositing, offset, Offset.zero & size, super.paint)` ?

Reply:
Because that causes problems.  It doesn't offset the child based on the alignment calculations, and so when the overflow starts, the child pops to 0,0.